### PR TITLE
UIREQ-1119 - Implement availability of "Print status" filters upon "Enable view print details (Pick slips)" configuration.

### PR DESCRIPTION
**Pull request by manvendra-s-rathore**
[UIREQ-1119](https://folio-org.atlassian.net/browse/UIREQ-1119) - Implement availability of "Print status" filters upon "Enable view print details (Pick slips)" configuration.

## Scenarios 
- When  ‘Enable view print details (Pick slips)’ setting is enabled.
It should display 'Print status' Filters and filter request records on basis of '# Copies' column. 
When any of 'Print Status' Filter is selected, it should export only filtered records in Action Menu > 'Export search results to csv'.
- When  ‘Enable view print details (Pick slips)’ setting is disabled.
It should not display 'Print Status' Filters.

## Screencast
- When  ‘Enable view print details (Pick slips)’ setting is enabled.

https://github.com/user-attachments/assets/cd61b831-b330-4cce-a325-618da5d42e00

- When  ‘Enable view print details (Pick slips)’ setting is disabled.

https://github.com/user-attachments/assets/240cae03-f542-4df3-97b2-7d895399f74c

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [x] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [x] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/components/RequestsFilters/RequestsFilters.js [hunk 1]
```diff
@@ -141,6 +146,26 @@ export default class RequestsFilters extends React.Component {
...
+            onClearFilter={() => onClear(requestFilterTypes.PRINT_STATUS)}
+          >
+            <CheckboxFilter
+              data-testid={`${requestFilterTypes.PRINT_STATUS}Filter`}
+              dataOptions={this.transformRequestFilterOptions(requestPrintStatusFilters)}
+              name={requestFilterTypes.PRINT_STATUS}
+              selectedValues={printStatus}
+              onChange={onChange}
+            />
+          </Accordion>
```
Comment (Terala-Priyanka) [inline src/components/RequestsFilters/RequestsFilters.js:167]: Could you please move this logic to another component?

File: src/components/RequestsFilters/RequestsFilters.test.js [hunk 2]
```diff
@@ -317,4 +322,60 @@ describe('RequestsFilters', () => {
       expect(RequestLevelFilter).toHaveBeenCalledWith(expect.objectContaining(expectedProps), {});
     });
   });
+
+  describe('Print status accordion', () => {
```
Comment (Terala-Priyanka) [inline src/components/RequestsFilters/RequestsFilters.test.js:326]: Could you please add tests for the behavior 'when isViewPrintDetailsEnabled is not enabled'

Comment (manvendra-s-rathore) [reply src/components/RequestsFilters/RequestsFilters.test.js:326]: test case for the behavior 'when isViewPrintDetailsEnabled is not enabled' is covered already. I will move it to separate describe for better readability.
image

File: src/routes/utils.js [hunk 3]
```diff
@@ -41,3 +44,23 @@ export const getFullNameForCsvRecords = (record) => {
   const { firstName = '', middleName = '', lastName = '' } = record;
   return [firstName, middleName, lastName].filter(Boolean).join(' ');
 };
+
+export const filterRecordsByPrintStatus = (records, printStatusFilters) => {
+  const isPrintedFilterSelected = printStatusFilters[0] === requestPrintStatusType.PRINTED;
+  return records.filter(record => {
+    const hasPrintDetails = record.printDetails != null;
```
Comment (Terala-Priyanka) [inline src/routes/utils.js:51]: Could you please check if this is needed?

File: src/routes/RequestsRoute.js [hunk 4]
```diff
@@ -1388,6 +1408,8 @@ class RequestsRoute extends React.Component {
     const pickSlipsData = convertToSlipData(pickSlips, intl, timezone, locale, SLIPS_TYPE.PICK_SLIP, user);
     const searchSlipsData = convertToSlipData(searchSlips, intl, timezone, locale, SLIPS_TYPE.SEARCH_SLIP_HOLD_REQUESTS);
     let multiSelectPickSlipData = getSelectedSlipDataMulti(pickSlipsData, selectedRows);
+    const displayPrintStatusFilteredData = isViewPrintDetailsEnabled &&
+      resources.records.hasLoaded && selectedPrintStatusFilters.length === 1;
```
Comment (Terala-Priyanka) [inline src/routes/RequestsRoute.js:1412]: Please add necessary comment here.

File: src/routes/RequestsRoute.js [hunk 5]
```diff
@@ -673,26 +677,34 @@ class RequestsRoute extends React.Component {
...
       }
     }
 
+    if (isViewPrintDetailsEnabled && selectedPrintStatusFilters.length === 0 && filters) {
+      const printStatusFilterInQuery = this.getActiveFilters()[requestFilterTypes.PRINT_STATUS];
+
+      if (printStatusFilterInQuery && printStatusFilterInQuery.length === 1) {
+        this.setState({ selectedPrintStatusFilters: [printStatusFilterInQuery[0]] });
+      }
+    }
```
Comment (Terala-Priyanka) [inline src/routes/RequestsRoute.js:706]: Try to refactor this code to bring down the cognitive complexity as explained by Sonar

File: src/routes/RequestsRoute.js [hunk 6]
```diff
@@ -673,26 +677,34 @@ class RequestsRoute extends React.Component {
...
+      const isPrintDetailsSettingsEnabled = currPrintDetailsSettings === 'true';
+      this.setState({ isViewPrintDetailsEnabled: isPrintDetailsSettingsEnabled });
 
-      if (!isViewPrintDetailsEnabled) {
+      if (!isPrintDetailsSettingsEnabled) {
         this.columnHeadersMap = getFilteredColumnHeadersMap(this.columnHeadersMap);
       }
     }
 
+    if (isViewPrintDetailsEnabled && selectedPrintStatusFilters.length === 0 && filters) {
```
Comment (Terala-Priyanka) [inline src/routes/RequestsRoute.js:700]: `isViewPrintDetailsEnabled` may not point to latest/updated state. Please check.
Here there are couple of variables - `currPrintDetailsSettings ` and `isViewPrintDetailsEnabled` referring the same value. It needs to be addressed.

**Review by Terala-Priyanka — CHANGES_REQUESTED** [review]
Please address the comments.

File: src/routes/RequestsRoute.js [hunk 7]
```diff
@@ -731,6 +733,34 @@ class RequestsRoute extends React.Component {
     if (!this.props.resources.records.isPending) {
       this.onSearchComplete(this.props.resources.records);
     }
+
+    if (isViewPrintDetailsEnabled !== prevState.isViewPrintDetailsEnabled && !isViewPrintDetailsEnabled) {
+      this.columnHeadersMap = getFilteredColumnHeadersMap(this.columnHeadersMap);
+    }
+
+    this.updateSelectedPrintStatusFilters(isViewPrintDetailsEnabled, selectedPrintStatusFilters, filters);
```
Comment (Terala-Priyanka) [inline src/routes/RequestsRoute.js:741]: Please check the check the presence of "print status" filter selection within the "filters" that you receive from query props, if needed call updateSelectedPrintStatusFilters method accordingly.

**Review by Terala-Priyanka — CHANGES_REQUESTED** [review]

File: src/routes/RequestsRoute.js (hunk 7 above)
Comment (manvendra-s-rathore) [reply src/routes/RequestsRoute.js]: yes, refactored and updated. Thanks.

**Review by SKarolFolio — APPROVED** [review]

**Review by Terala-Priyanka — APPROVED** [review]

**Pull request merged**